Vizzuality/trase
# Feature/streaming downloads

## Pivotal Tracker

https://www.pivotaltracker.com/story/show/170797946

## Description

Ended up not optimising transposing (pivoting) downloads, as it seemed there's not much to be gained without changing this functionality completely.

Instead I applied a different optimisation, related to how files are served to the client. In particular:
- precomputed files used to be sent using `send_data`, now `send_file`. I think that will make it easier to apply [accelerated downloads with X-Accel-Redirect](https://mattbrictson.com/accelerated-rails-downloads) once we move the whole lot from Apache to NGINX
- generated files are now streamed. To achieve that, we use the `zip_tricks` gem, which combined with `ActionController::Live` allows to stream zip files on the fly. This is particularly useful in case of bigger datasets, where we split the files by year. Once a file for a given year has been generated, it can already be streamed. The advantage is lower memory usage and better user experience, as it becomes more obvious that data is coming, even if still rather slowly.

## Testing instructions

- Download a custom download (e.g. select Brazil soy, all years but one - that will be generated rather than precomputed) and check that the client can now display the download progress

## Deploy instructions

Varnish config: set `between_bytes_timeout` (300s) and test with Brazil beef
